Implemented `no_std` support (with/without `alloc`)

JRRudy1 opened the pull request:
This PR adds support for using the `transient` crate in `no_std` environments, with or without `alloc`, through the addition of new "std" and "alloc" crate features. This PR also includes CI updates that run the tests with various feature sets to ensure everything works regardless of the chosen crate features.

To enable `no_std` mode, disable default features using `--no-default-features` (or `default-features = false` in Cargo.toml). Almost all functionality of the `transient` crate will still be available, except for:
- the `downcast`/`downcast_unchecked` methods on the `dyn Any` trait object (which accepts/returns a `Box`)
- the `Transient::erase` convenience method (which accepts/returns a `Box`)
- `Transient` impls for non-`core` types (particularly `Box`, `Vec`, `String`, and `HashMap`) 

To enable `no_std` mode with `alloc` support, disable default features and enable the "alloc" feature using  `--no-default-features --features alloc` (or `default-features = false, features = ["alloc"]` in Cargo.toml). This restores all of the functionality listed above except for the `Transient` impls for several types that are available in neither `core` nor `alloc`.

Resolves #25

JRRudy1 committed:
Added new "std" and "alloc" crate features to Cargo.toml

JRRudy1 committed:
Split the stdlib `Transient` impls into submodules by their dependence on `std` and `alloc`, and updated `use` statements accordingly.

JRRudy1 committed:
Changed `PhantomData` import source from `std` to `core`.

JRRudy1 committed:
Changed `std` imports to `core` where possible, and added `std`/`alloc` feature gates elsewhere.

This also included reorganizing the `any` module tests a bit to reduce the number of feature gates needed.

JRRudy1 committed:
Updated tests for `no_std` support.

JRRudy1 committed:
Added `no_std` crate attribute gated by the "std" feature, and `extern use alloc` statement gated by the "alloc" feature.

This completes the implementation of `no_std` support.

JRRudy1 committed:
Added CI matrix dimension for the set of crate features to enable during tests.

JRRudy1 committed:
Removed the "--all" arg from the CI build step to avoid building the derive crate in the test matrix.

JRRudy1 committed:
Added the "std" feature as a dependency of the "derive" feature.

JRRudy1 committed:
Added feature-gates to several doctests to avoid errors when running tests with different feature sets.

JRRudy1 committed:
rustfmt

JRRudy1 committed:
Updated the `Transient` impl for `AddrParseError` to require the "std" feature on Rust versions before the type was moved from `std` to `core`.

JRRudy1 committed:
Reordered test args in CI workflow for better display on Github.

JRRudy1 committed:
Formatting fix

JRRudy1 commented:
@the10thWiz you may want to take a peak at this; Rocket obviously doesn't need `no_std` and the changes should be completely transparent, but it does slightly complicate development since `std` usages will need to be either replaced by `core` if possible or gated by `cfg(feature = "std")` if not

JRRudy1 committed:
Simplified conditional imports by resolving them once in a facade module and then importing from it instead of `std` or `alloc`.

JRRudy1 committed:
Fixed missing feature gate and added no_std support to the crate documentation.

JRRudy1 committed:
Fixed ambiguous import by adding an explicit `::` prefix.

the10thWiz committed:
Remove wrangling since `std` types are just re-exports of `core` and `alloc` types whenever possible

the10thWiz committed:
Fix remaining uses in tests

the10thWiz committed:
Fix formatting

JRRudy1 commented:
@the10thWiz simplified the implementation quite a bit in #28, and the changes have been pulled into this PR

JRRudy1 committed:
Cleanup

the10thWiz commented on Cargo.toml:
```diff
@@ -35,10 +35,19 @@ rmp-serde = { version = ">=1", default-features = false, optional = true }
 
 # Provides a `derive` macro for implementing the `Transient` trait
-derive = ["transient-derive"]
+derive = ["std", "transient-derive"]
```
The derive macro does not rely on `std`, so this shouldn't be needed.
```suggestion
derive = ["transient-derive"]
```

JRRudy1 replied:
My logic on adding that was mostly to communicate that `transient-derive` (and proc-macro crates in general by my understanding) is *not* `no_std`, and that implicitly pulling it in by leaving the `derive` feature active will require `std`. So disabling the "std" feature but keeping the "derive" feature would be pretty pointless.

However, I see that Serde follows your logic instead of mine, so maybe that is the way to go. What do you think?

the10thWiz replied:
Executing the derive macro requires std *at build time*. Proc-macro crates do not implicitly add a dependency on std. A `no_std` crate can still use derive macros, as long as the generated code doesn't use anything from std.

In fact, one obvious example is serde - `no_std` crates can use serde's derive macros.

JRRudy1 replied:
Wow, I have never had to deal with `no_std` environments in my work and didn't even realize you could use derives in them! Very good to know, thanks for the review!

the10thWiz reviewed (commented):
This looks pretty great to me. I just have one suggestion.

JRRudy1 committed:
Removed the "std" feature as a dependency of the "derive" feature.

As pointed out by @the10thWiz, this dependency was not necessary and removing it adds flexibility.

Co-authored-by: Matthew Pomes <matthew.pomes@pm.me>
